packages/pyright-internal/src/analyzer/cacheManager.ts

```diff
@@ -24,12 +27,41 @@ export interface CacheOwner {
 export class CacheManager {
     private _pausedCount = 0;
     private readonly _cacheOwners: CacheOwner[] = [];
+    private _sharedUsageBuffer: SharedArrayBuffer | undefined;
+    private _sharedUsagePosition = 0;
     private _lastHeapStats = Date.now();
 
+    constructor(private readonly _maxWorkers: number = 0) {}
     registerCacheOwner(provider: CacheOwner) {
         this._cacheOwners.push(provider);
     }
 
+    addWorker(index: number, worker: Worker) {
+        // Send the sharedArrayBuffer to the worker so it can be used
+        // to keep track of heap usage on all threads.
+        const buffer = this._getSharedUsageBuffer();
+        if (buffer) {
+            // The SharedArrayBuffer needs to be separate from data in order for it
+            // to be marshalled correctly.
+            worker.postMessage({ requestType: 'cacheUsageBuffer', sharedUsageBuffer: buffer, data: index.toString() });
+            worker.on('exit', () => {
+                const view = new Float64Array(buffer);
+                view[index] = 0;
+            });
+        }
+    }
+
+    handleCachedUsageBufferMessage(msg: AnalysisRequest) {
+        if (msg.requestType === 'cacheUsageBuffer') {
+            const index = parseInt(msg.data || '0');
+            const buffer = msg.sharedUsageBuffer;
+            if (buffer && index) {
```

Comment:
Isn't 0 a legal index? Or is an index of 0 reserved for the main thread?

This `if` statement will evaluate to falsy if index is 0.

Reply:
0 is the index reserved for the main thread. I'll add a comment.